### ISSUE TITLE
BUG: Leaking go routines in kubelet

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#19633]
In testing cpu performance as part of #19341 , I uncovered leaking go routines from within cadvisor on the kubelet.  

---
Stack 1:  Worst case
fedora 23 
tip of master 
golang 1.5.2 
etcd 2.2.1 
docker 1.9.1 

```
$ sudo ./hack/local-up-cluster.sh 
$ submit single pause with replica of 1
<measure pprof goroutine data on kubelet as the initial control>
```
wait till running
```
$ kubectl scale --replicas=50 rc/wide-pause
<wait till kubelets starts pods> 
$ kubectl scale --replicas=1 rc/wide-pause
<wait till kubelets starts terminating pods> 
$  kubectl scale --replicas=100 rc/wide-pause
<notice overlap, total pod count over 100> 
```
The system becomes remarkably slow and takes > 20 minutes to fill! repros in other condition, see below.
```
$ kubectl scale --replicas=1 rc/wide-pause
<measure pprof goroutine again and check for changes, and repeat experiment> 
````

recorded iteration1 (with only 1 pod running): 
336 @ (*realFsHandler).trackUsage+0x482	/home/tsclair/work/spaces/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/google/cadvisor/container/docker/fsHandler.go:82

recorded iteration2 (with only 1 pod running): 
528 @ github.com/google/cadvisor/container/docker.(*realFsHandler).trackUsage+0x482	/home/tsclair/work/spaces/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/google/cadvisor/container/docker/fsHandler.go:82

**LEAK** 

---
I can reproduce a much smaller leak using a production EL stack

Stack 2: 
rhel7.2
tip of master 
golang 1.4.2 
etcd 2.1.1 
docker 1.9.1 

same procedure: 
iteration1: 
 4 @ github.com/google/cadvisor/container/docker.(*realFsHandler).trackUsage+0x3ff	/home/cloud-user/kubernetes/Godeps/_workspace/src/github.com/google/cadvisor/container/docker/fsHandler.go:82

iteration2:
10 @ github.com/google/cadvisor/container/docker.(*realFsHandler).trackUsage+0x3ff	/home/cloud-user/kubernetes/Godeps/_workspace/src/github.com/google/cadvisor/container/docker/fsHandler.go:82

iteration3:
14 @ github.com/google/cadvisor/container/docker.(*realFsHandler).trackUsage+0x3ff	/home/cloud-user/kubernetes/Godeps/_workspace/src/github.com/google/cadvisor/container/docker/fsHandler.go:82

---

I'm not entirely certain why the fedora case is so much worse.  Either golang 1.5, or possibly some changes in systemd/cgroups which is causing cadvisor to loose it.

/cc @kubernetes/sig-node @mattf @jeremyeder @jimmidyson

**Comment by jimmidyson** [comment]
I've recreated this with cadvisor at git commit google/cadvisor@634965abc45557ed03c268bb193a00cfcbedbd32 vendored in Kubernetes as well as cadvisor master. Seems like cadvisor starts too many goroutines when multiple containers are started in parallel. Sequentially starting/stopping containers works fine & goroutines are properly cleaned up. Looking for root cause.

**Comment by jimmidyson** [comment]
Tracked it down - fixing now. Leak is a race, unsurprisingly.

Patching cadvisor now, will send in a PR to kubernetes once it's merged & release tagged.

**Comment by jeremyeder** [comment]
Awesome, thanks guys!  Indeed, as you can imagine we are hitting races all over the stack (kernel included) when we dial-up the input rates.

**Comment by timothysc** [comment]
@jimmidyson @vishh next release of cadvisor we should update our godeps asap.  

/cc @ncdc apply 1.2-candidate please.

**Comment by jimmidyson** [comment]
@timothysc Absolutely - I'll sort that out ASAP. I have another issue that we need to update cadvisor to fix as well which is waiting on a tag being applied even further upstream (opecontainers/runc)... These would both be required for 1.2 so would rather do the Kubernetes PR to update cadvisor once.

**Comment by timothysc** [comment]
:+1:

**Comment by gmarek** [comment]
cc @kubernetes/sig-scalability @wojtek-t @gmarek

**Comment by timothysc** [comment]
I'll open a separate issue on the speed of overlapping operations.

**Comment by timothysc** [comment]
Fix for the leakage has been pushed to the cadvisor repo, once the godep has been updated we can close.

**Comment by timothysc** [comment]
Should be fixed by: https://github.com/kubernetes/kubernetes/pull/19739 once in, we can close 

/cc @vishh
